Use LocaleProvider in calypso

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Implements usage of LocaleProvider, useLocalizeUrl
* Implements withLocalizeUrl 
* Makes use of both with/useLocalizeUrl
* adds i18n-utils to domain-picker package

#### Testing instructions

* Check url is localized as shown in #45610
* Check urls is localized on the privacy settings screen
* Check existing localized urls still localize elsewhere in calypso
 
Fixes #45610 :tada:

lsl committed:
Apply LocaleProvider to calypso

lsl committed:
Implement localizeUrl usage in calspyo with LocaleProvider

lsl committed:
Remove debugging

p-jackson commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -25,7 +27,11 @@ const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
 
-	return <I18nProvider localeData={ localeData }>{ children }</I18nProvider>;
+	return (
+		<LocaleProvider localeSlug={ localeSlug || i18nDefaultLocaleSlug }>
```
Because `<CalypsoI18nProvider>` is very app specific I'd use the `i18n_default_locale_slug` config instead of the default that comes from the package.

Pretty nit picky given it's the same value and `i18n_default_locale_slug` will likely never change.

```suggestion
		<LocaleProvider localeSlug={ localeSlug || config( 'i18n_default_locale_slug' ) }>
```

lsl replied:
Happy to use this for now but realistically the config values need to be removed now that they are required outside of calypso package. Why I moved all of them to hard coded values in i18n-utils package.

p-jackson reviewed (commented):
`domain-picker` is also used in the ETK's launch flow it will need a `<LocaleProvider>` added somewhere near the root. Probably in `attach-launch-sidebar.tsx`

p-jackson commented:
Lots of CI errors, rebased in case it needed it after the `trunk` branch rename.

p-jackson commented:
I'm at a bit of a loss as to what the build error's about. I see the error in the local build logs too, but I just can't see what the problem is.

lsl committed:
Update client/components/calypso-i18n-provider/index.tsx

Co-authored-by: Philip Jackson <p-jackson@live.com>

p-jackson committed:
Infer `withLocalizeUrl`'s type

lsl commented:
> Lots of CI errors, rebased in case it needed it after the trunk branch rename.

I don't think this took, tried a fresh rebase. Should probably retest, the initial testing was done against master that was probably a few days out of date.

lsl commented:
That didn't work, testing moving to fresh PR in case its something build related https://github.com/Automattic/wp-calypso/pull/47692

p-jackson commented:
Maybe it's something to do with the filename changing from `.ts` to `.tsx`?

p-jackson commented:
Actually I wonder if this is something to do build order.
If I do `yarn clean && yarn` to clean the `dist` folders and build the packages then I get a build error. Then if I follow up with another `yarn` then the build succeeds. The `domain-picker` package is complaining about `i18n-utils` not existing. But after starting the second build the `i18n-utils/dist/types` folder exists and now. Maybe the order packages are built is alphabetical so lerna builds `domain-picker` first, followed by `i18n-utils`. This might just be the first time that this has hit us.

Looking up lerna docs to see if there's a way to change the order workspace operations are run in.

p-jackson committed:
Change package build order to fix build

p-jackson commented:
@scinos we had a broken build here before the `domain-picker` package was being built before `i18n-utils`, even though `i18n-utils` depends on `domain-picker`. Pretty unfortunate if we need to main this build ordering manually :(

[From this comment](https://github.com/microsoft/TypeScript/issues/29643#issuecomment-459187348) it sounds like it's possible to have `tsc` build packages in the correct order if you also maintain a `references` list in each of the sibling package's `tsconfig.json` too. Which would be a lot of duplication.

[TypeScript has a giant tracking issue](https://github.com/microsoft/TypeScript/issues/25376) to support this without having to maintain another list of dependencies.

scinos commented:
Looks like adding the reference to `i18n-utils` in `packages/domain-picker/tsconfig.json` fixed it (in https://github.com/Automattic/wp-calypso/pull/47692)

I agree it is not great to have to declare the dependencies in `package.json` and `tsconfig.json`, but I don't see how we can improve it until https://github.com/microsoft/TypeScript/issues/25376 is resolved.

Maybe we can start by having some `tsconfig.json` linting that ensures all links are there? At least we would be able to produce a clear error, not something obscure that takes time to diagnose.

p-jackson committed:
Get default locale slug from i18n-utils, not config

p-jackson committed:
Update LocaleProvider when locale changes

p-jackson commented:
Updated so that the locale slug in context updates when when `i18n-calypso` fires the `'change'` event. Also fixed lint error and force pushed to bring things up to date.

p-jackson committed:
Add TS reference to @automattic/i18n-utils

Don't need to have maintain a specific build order for packages if the
domain-picker package just declares it's dependency on i18n-utils

p-jackson commented:
> Looks like adding the reference to `i18n-utils` in `packages/domain-picker/tsconfig.json` fixed it

Awesome! I've brought that fix into this PR and closed #47692 since it was more just a trial PR to see if we were dealing with some GH or CI issue.

p-jackson reviewed (approved):
Tested and everything to do with the change is working as expected.

I tested the domain picker inside the launch flow too (so inside the ETK). The HSTS tooltip doesn't appear when you click the button. It's a bug that needs to be fixed, but not i18n related.

The woo e2e tests are failing, but I think it's because there's currently something wrong with the e2e test user's site. When I run it locally the user's dashboard looks like this:
<img width="966" alt="Screenshot 2020-11-30 at 5 14 43 PM" src="https://user-images.githubusercontent.com/1500769/100568287-b8181e80-332f-11eb-9938-28de6fac1731.png">

Maybe the e2e tests have just caught the site at a bad time 🤷

p-jackson commented:
Re-ran tests, passing now. Whatever was afflicting the woo test user must have been fixed.

lsl commented:
Rebased, retested, everything seems ok.
